from PyQt5.QtWidgets import *
from PyQt5.QtCore import *
from PyQt5.QtGui import *
import threading
import sys
# å¸®æˆ‘ä»¬ç›´æ¥è¿è¡Œæ­¤æ–‡ä»¶æ—¶ï¼Œå¯ä»¥åŠ è½½åˆ°ä¸Šçº§ç›®å½•çš„uiåŒ…
sys.path.append("../")

from common.utils import decode_data
from drivers.driver_bluetooth import BluetoothDataTransfer
from ui.Ui_widget_bluetooth import Ui_BluetoothWidget

"""
PyQtä¸­è‡ªå®šä¹‰ä¿¡å·å’Œæ§½çš„æ­¥éª¤å¦‚ä¸‹ï¼š

1.  å®šä¹‰ä¿¡å·ï¼šåœ¨ç±»ä¸­å®šä¹‰ä¸€ä¸ªä¿¡å·ï¼Œä½¿ç”¨PyQtçš„ä¿¡å·æœºåˆ¶å¯ä»¥å®ç°è‡ªå®šä¹‰ä¿¡å·ã€‚å¯ä»¥é€šè¿‡pyqtSignal()æ–¹æ³•æ¥åˆ›å»ºä¸€ä¸ªä¿¡å·å¯¹è±¡ã€‚pyqtSignalå‚æ•°æ§½å‡½æ•°çš„å‚æ•°ç±»å‹
2.  å®šä¹‰æ§½ï¼šåœ¨ç±»ä¸­å®šä¹‰ä¸€ä¸ªæ§½å‡½æ•°ï¼Œä½¿ç”¨@pyqtSlot()è£…é¥°å™¨å°†è¯¥æ–¹æ³•æ³¨å†Œä¸ºæ§½å‡½æ•°ã€‚ 
3.  è¿æ¥ä¿¡å·å’Œæ§½ï¼šä½¿ç”¨ä¿¡å·.connect(æ§½å‡½æ•°)æ–¹æ³•å°†ä¿¡å·å’Œæ§½è¿æ¥èµ·æ¥ã€‚ 
4.  è§¦å‘ä¿¡å·ï¼šåœ¨éœ€è¦è§¦å‘ä¿¡å·çš„åœ°æ–¹ï¼Œä½¿ç”¨ä¿¡å·.emit()æ–¹æ³•æ¥è§¦å‘è¯¥ä¿¡å·ï¼Œå‚æ•°å’Œæ§½å‡½æ•°å‚æ•°ä¸€è‡´ã€‚
"""


class BluetoothWidget(QWidget):
    
    # 1. å®šä¹‰ä¿¡å·ï¼š
    scan_device_signal = pyqtSignal(list)

    def __init__(self, parent=None):
        super().__init__(parent)
        self.main_window : QMainWindow = parent
        self.ui = Ui_BluetoothWidget()
        self.ui.setupUi(self)
        
        self.current_bdt: BluetoothDataTransfer = None
        # 3.  è¿æ¥ä¿¡å·å’Œæ§½ï¼š
        self.scan_device_signal.connect(self.on_scan_device_result)

        self.init_ui()
        
        # é»˜è®¤éšè—
        self.ui.gb_connected_device.setVisible(False)
        
    def __show_status(self, msg, msecs=0):
        if self.main_window is None:
            return
        
        # ç¡®ä¿parentæ˜¯QMainWindow
        if isinstance(self.main_window, QMainWindow):
            self.main_window.statusBar().showMessage(msg, msecs)
            
    # def on_btn_connect_clicked(self, arg=""):
    #     print("è¿æ¥ğŸ”—", arg)
    
    def refresh_devices(self):
        print("åˆ·æ–°ğŸ”ƒ", threading.current_thread().name)
        devices = BluetoothDataTransfer.scan_devices()
        # 4.  è§¦å‘ä¿¡å·ï¼š
        # self.on_scan_device_result(devices)
        self.scan_device_signal.emit(devices)
        
    # 2.  å®šä¹‰æ§½ï¼š
    @pyqtSlot(list)
    def on_scan_device_result(self, devices: list):
        print("æ›´æ–°UIçº¿ç¨‹:", threading.current_thread().name) 
        
        # pyqt5:å»ºè®®åœ¨ä¸»çº¿ç¨‹æ›´æ–°UI
        
        self.__show_status(f"æ‰«æåˆ°{len(devices)}ä¸ªè®¾å¤‡")
        
        # å¯ç”¨åˆ·æ–°æŒ‰é’®
        self.ui.btn_refresh.setEnabled(True)
        
        # æŠŠæ‰«æåˆ°çš„è®¾å¤‡åŠ å…¥åˆ°ä¸‹æ‹‰åˆ—è¡¨ä¸­
        self.ui.cb_devices.clear()
        # å¾ªç¯æ‰€æœ‰è®¾å¤‡
        for device in devices:
            # addr, name = device
            self.ui.cb_devices.addItem(device[1], device)
            
        # é»˜è®¤é€‰ä¸­ç¬¬ä¸€ä¸ªè®¾å¤‡
        if self.ui.cb_devices.count() > 0:
            self.ui.cb_devices.setCurrentIndex(0)
    
    def on_refresh_clicked(self):
        # ç¦ç”¨æ‰åˆ·æ–°æŒ‰é’®ï¼Œé¿å…é‡å¤è¿›å…¥åˆ·æ–°
        self.ui.btn_refresh.setEnabled(False)
        
        # rst = threading.Thread(target=BluetoothDataTransfer.scan_devices).start()
        thread = threading.Thread(target=self.refresh_devices, daemon=True)
        thread.start()
        
    
    def on_connect_clicked(self):
        if self.current_bdt is not None:
            print("æ–­å¼€è¿æ¥ï¼š", self.current_bdt.target_name)
            self.current_bdt.disconnect()
            self.__show_status(f"å·²æ–­å¼€{self.current_bdt.target_name}çš„è¿æ¥", 2000)
            self.current_bdt = None
            self.ui.btn_connect.setText("è¿æ¥ğŸ”—")
            self.ui.label_state.setPixmap(QPixmap(":/icon/disconnect"))
            self.ui.gb_connected_device.setVisible(False)
            return
            
        print("è¿æ¥ğŸ”—")
        device = self.ui.cb_devices.currentData()
        if device is None:
            print("è¯·å…ˆæ‰«æå¹¶é€‰æ‹©è®¾å¤‡ï¼")
            QMessageBox.warning(self, "è­¦å‘Š", "è¯·å…ˆæ‰«æå¹¶é€‰æ‹©è®¾å¤‡ï¼", QMessageBox.Yes)
            return
        
        addr, name = device
        bdt = BluetoothDataTransfer(addr, name)
        # bdt = BluetoothDataTransfer(*device)
        if bdt.connect():
            self.ui.btn_connect.setText("æ–­å¼€ğŸ”Œ")
            self.ui.label_state.setPixmap(QPixmap(":/icon/connect"))
            self.ui.label_addr.setText(f"åç§°:{name}")
            self.ui.label_name.setText(f"åœ°å€:{addr}")
            self.ui.gb_connected_device.setVisible(True) # ä¿®æ”¹å¯è§çŠ¶æ€
            self.__show_status(f"å·²è¿æ¥{name}çš„è®¾å¤‡")
            
            self.current_bdt = bdt
            
            thread = threading.Thread(target=self.recv_msg_loop, daemon=True)
            thread.start()
            
    def recv_msg_loop(self):
        while True:
            # å¦‚æœè®¾å¤‡å·²æ–­å¼€, åˆ™é€€å‡ºå¾ªç¯
            if self.current_bdt is None:
                break
            
            data = self.current_bdt.receive_data()
            if data:
                print("data->", data)
                self.ui.edit_recv.append(decode_data(data))
            else:
                # å¯¹æ–¹å…³é—­è¿æ¥, é€€å‡ºå¾ªç¯
                break
            
    @pyqtSlot()
    def on_send_clicked(self):
        # æ£€æŸ¥æ˜¯å¦æœ‰è¿æ¥çš„è®¾å¤‡
        if self.current_bdt is None:
            print("è¯·å…ˆè¿æ¥è®¾å¤‡ï¼")
            QMessageBox.warning(self, "è­¦å‘Š", "è¯·å…ˆè¿æ¥è®¾å¤‡ï¼", QMessageBox.Yes)
            return
        
        # è·å–è¾“å…¥æ¡†çš„å†…å®¹ï¼ˆåˆ¤å®šæ˜¯å¦è¾“å…¥äº†å†…å®¹ï¼‰
        text = self.ui.edit_send.toPlainText()
        if text == "":
            print("è¯·è¾“å…¥è¦å‘é€çš„å†…å®¹ï¼")
            QMessageBox.warning(self, "è­¦å‘Š", "è¯·è¾“å…¥è¦å‘é€çš„å†…å®¹ï¼", QMessageBox.Yes)
            return
        
        # æ”¯æŒç›´æ¥å‘é€å­—ç¬¦ä¸²(é»˜è®¤UTF-8ç¼–ç ), ä¹Ÿå¯ä»¥å‘é€å­—èŠ‚æµ(å­—èŠ‚æ•°ç»„)
        # self.current_bdt.send_data(text.encode())
        # self.current_bdt.send_data(b"\xAF\x13")
        self.current_bdt.send_data(text.encode("gbk"))

    def init_ui(self):
        self.ui.btn_connect.clicked.connect(self.on_connect_clicked)
        self.ui.btn_refresh.clicked.connect(self.on_refresh_clicked)
        self.ui.btn_send.clicked.connect(self.on_send_clicked)
        self.ui.btn_send_clear.clicked.connect(self.ui.edit_send.clear)
        self.ui.btn_recv_clear.clicked.connect(self.ui.edit_recv.clear)


if __name__ == '__main__':
    app = QApplication(sys.argv)

    window = BluetoothWidget()
    window.show()

    sys.exit(app.exec_())